src/Workspaces/CSharp/Portable/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs

```diff
@@ -0,0 +1,289 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.EmbeddedLanguages.VirtualChars
+{
+    [ExportLanguageService(typeof(IVirtualCharService), LanguageNames.CSharp), Shared]
+    internal class CSharpVirtualCharService : AbstractVirtualCharService
+    {
+        public static readonly IVirtualCharService Instance = new CSharpVirtualCharService();
+
+        protected override bool IsStringLiteralToken(SyntaxToken token)
+            => token.Kind() == SyntaxKind.StringLiteralToken;
+
+        protected override ImmutableArray<VirtualChar> TryConvertToVirtualCharsWorker(SyntaxToken token)
+        {
+            Debug.Assert(!token.ContainsDiagnostics);
+            if (token.Kind() == SyntaxKind.StringLiteralToken)
+            {
+                return token.IsVerbatimStringLiteral()
+                    ? TryConvertVerbatimStringToVirtualChars(token, "@\"", "\"", escapeBraces: false)
+                    : TryConvertStringToVirtualChars(token, "\"", "\"", escapeBraces: false);
+            }
+            else if (token.Kind() == SyntaxKind.InterpolatedStringTextToken)
+            {
+                var interpolatedString = (InterpolatedStringExpressionSyntax)token.Parent.Parent;
+                return interpolatedString.StringStartToken.Kind() == SyntaxKind.InterpolatedVerbatimStringStartToken
+                    ? TryConvertVerbatimStringToVirtualChars(token, "", "", escapeBraces: true)
+                    : TryConvertStringToVirtualChars(token, "", "", escapeBraces: true);
+            }
+            else
+            {
+                return default;
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

I filed [https://github.com/dotnet/roslyn/issues/29172](https://github.com/dotnet/roslyn/issues/29172) to follow-up when the alternate interpolated verbatim strings feature (@$" instead of $@") gets merged. This code may have to be updated.

Reply:
Could you add a test to `CSharpVirtualCharServiceTests.cs` that `@$"` produces specific diagnostics? Those diagnostics will change when C# 8 is merged in and that can be used as a reminder. Thanks

---
In reply to: [208761728](https://github.com/dotnet/roslyn/pull/28927#discussion_r208761728) [](ancestors = 208761728)

Reply:
sure.

Reply:
test added.